### PR TITLE
CORE beacon scalability

### PR DESCRIPTION
**Pull request by tzaeschke**
Discussion on core beacon scalability and Recovery

Contributes to #2

### CODE REVIEW

File: draft-meynell-panrg-scion-research-questions.md [hunk 1]
```diff
@@ -200,6 +200,26 @@ This draft assumes the reader is familiar with some of the fundamental concepts
...
+## CORE Beaconing Scalability
+
+The idea behind beaconing is to discover all possible paths (loop free and with a fixed maximu length) between two CORE AS.
+Every AS forwards any received becons to all neighbor CORE ASes unless this would cause a loop or exceed the fixed maximum length.
+Implemented naively, the number of paths (and beacons) grows superexponentially with the network size.
+This is currently mitigated primarily (and efficiently) by forwarding only the 5 "best PCB" to neighbours.
+
+This "best PCB" practive has several implications:
+
+* It limits the number of available paths for endhosts such that only 5 path may be available to a given destination. This is especially relevant for
```
[the remark this answers is not in the document]
Comment (tzaeschke) [reply draft-meynell-panrg-scion-research-questions.md:212]: I think usually the same 5 paths would forwarded every time, if they are the best paths, then should it change and forward other paths? Or not?

File: draft-meynell-panrg-scion-research-questions.md [hunk 2]
```diff
@@ -200,6 +200,26 @@ This draft assumes the reader is familiar with some of the fundamental concepts
 * One AS belongs to many ISDs?
 
 
+## CORE Beaconing Scalability
+
+The idea behind beaconing is to discover all possible paths (loop free and with a fixed maximu length) between two CORE AS.
+Every AS forwards any received becons to all neighbor CORE ASes unless this would cause a loop or exceed the fixed maximum length.
+Implemented naively, the number of paths (and beacons) grows superexponentially with the network size.
```
[the remark this answers is not in the document]
Comment (nicorusti) [reply draft-meynell-panrg-scion-research-questions.md]: Two things: 
- I'd clarify the context for this number.. Is it in total in an overall SCION network? Is it per AS? I think it makes sense to talk about numbers per AS in general
- According to our (@matzf ) calculations and my understanding, if you limit the number of beacons with the "best PCB" practice, then the number is o(N*log(N)). So this would be if you were to not filter the number of beacons at all?

File: draft-meynell-panrg-scion-research-questions.md [hunk 3]
```diff
@@ -200,6 +200,26 @@ This draft assumes the reader is familiar with some of the fundamental concepts
...
+## CORE Beaconing Scalability
+
+The idea behind beaconing is to discover all possible paths (loop free and with a fixed maximu length) between two CORE AS.
+Every AS forwards any received becons to all neighbor CORE ASes unless this would cause a loop or exceed the fixed maximum length.
+Implemented naively, the number of paths (and beacons) grows exponentially with the network size.
+This is currently mitigated primarily (and efficiently) by forwarding only the 5 "best PCB" to neighbours.
+
+This "best PCB" practive has several implications:
+
+* It limits the number of available paths for endhosts such that only 5 path may be available to a given destination. This is especially relevant for
```
Comment (nicorusti) [inline draft-meynell-panrg-scion-research-questions.md:212]:
```suggestion
* It limits the number of available paths for endpoints such that only 5 path may be available to a given destination. This is especially relevant for
```

In drafts, we use `endpoint`  as per definition in RFC9473. Could you generally use endpoint instead of end host?

File: draft-meynell-panrg-scion-research-questions.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (nicorusti) [reply draft-meynell-panrg-scion-research-questions.md:212]: > such that only 5 path may be available to a given destination.  

This text gives the impression that there are at most 5 paths between a src and dst, which is not the case. Is this just an edge case? I think in general more paths are available between endpoints

File: draft-meynell-panrg-scion-research-questions.md [hunk 4]
```diff
@@ -200,6 +200,26 @@ This draft assumes the reader is familiar with some of the fundamental concepts
...
+  * Multipathing: you may not be able to use more than five path and no great selection of pats on offer .
+  * Unusual path policies, such as geofencing, may not be fulfillable by the limited number of paths on offer.
```
Comment (nicorusti) [inline draft-meynell-panrg-scion-research-questions.md:214]:
```
  * Unusual path policies, such as geofencing, may not be fulfillable by the limited number of paths on offer.
```
Was this ever an actual problem? Is it an edge case? I've never heard of this before

File: draft-meynell-panrg-scion-research-questions.md [hunk 5]
```diff
@@ -200,6 +200,26 @@ This draft assumes the reader is familiar with some of the fundamental concepts
...
+* It causes a small number of path to carry all traffic to a given destination, see {{link-load-balancing}}.
+* It opens the possibility of a DoS attack where a small number of "wormholes", see {{I-D.scion-cp}}, may be selected as "best" and have all traffic
```
Comment (nicorusti) [inline draft-meynell-panrg-scion-research-questions.md:216]: Are you referring to the `Wormhole Attack` we describe in the CP draft? 
Why is this related to core beaconing scalability?

File: draft-meynell-panrg-scion-research-questions.md [hunk 6]
```diff
@@ -276,6 +296,24 @@ When a client contacts a server, it is usually understood that it wants the serv
...
+
+* An endhost needs to tell it's local AS that the segment is bad
+* Local AS needs to tell CORE AS that segment is bad
+* The CORE AS needs to:
+  * change policy to exclude bad segments AND/OR
+  * tell other COREs and ISDs to stop delivering bad segment (they only deliver 5 each, so any bad segment shoud be avoided)
+
+Mitigation alternatives:
+
+* An AS could monitor traffic for SCMP errors, however this works only if these errors are actually generated and forwarded.
```
Comment (nicorusti) [inline draft-meynell-panrg-scion-research-questions.md:314]: Wouldn't a wormhole segment work? Is this section about segment revocation?

File: draft-meynell-panrg-scion-research-questions.md (hunk 4 above)
Comment (tzaeschke) [reply draft-meynell-panrg-scion-research-questions.md:215]: Not sure it happened yet, simply because we do not have much useful metadata yet (geofencing etc).
I think it can become a problem: if all ASes only forward their best five paths, then these five paths may be disjunct in themselves, but neighboring ASes are probably likely to forward the (almost) same 5 paths. In effect, an endpoint may have more than 5 paths available, but they are likely to be very similar and may not offer the choice that an endpoint desires?

File: draft-meynell-panrg-scion-research-questions.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (tzaeschke) [reply draft-meynell-panrg-scion-research-questions.md]: I added an explanation. Does it make sense? Otherwise I can remove the whole bullet point.

File: draft-meynell-panrg-scion-research-questions.md [hunk 7]
```diff
@@ -201,11 +201,36 @@ This draft assumes the reader is familiar with some of the fundamental concepts
...
+Every AS forwards any received beacons to all neighbor CORE ASes unless this would cause a loop or exceed the fixed maximum length.
+Implemented naively, the number of paths (and beacons) grows exponentially with the network size.
+This is currently mitigated, primarily (and efficiently), by forwarding only the five "best PCBs" to neighbours.
+
+This "best PCB" practice has several implications:
+
+* It limits the number of available paths for endpoints such that only five truly diverse paths may be available to a given destination.
+  "truly diverse" means: Most ASes will receive five PCBs to a given destination from several of their neighbours. However, assuming
+  those ASes are use a similar PCB selection policy, it is not unlikely that each neighbor AS will forward a very
+  similar set of five PCB. In effect, there may be more than five paths available, but they are likely to diverge only on the first
```
Comment (juagargi) [inline draft-meynell-panrg-scion-research-questions.md:216]: For this to happen, these neighbors have to share a common trailing path, e.g., being X our AS:
```
1) a->b->c -> X
2) a'->b'->c' -> X
etc.
```
Paths 1 and 2 would be "similar" if a' ≈ a and b'≈ b, while having c and c' being different neighbors.
Definitely not an impossible situation, but also not common: most likely the trailing path of neighbor c will be different that that of neighbor c', either in AS numbers or interface IDs.

File: draft-meynell-panrg-scion-research-questions.md [hunk 8]
```diff
@@ -201,11 +201,36 @@ This draft assumes the reader is familiar with some of the fundamental concepts
...
+## Beacon Forwarding Policy
+
+The idea behind "beaconing" is to discover all possible paths (loop free and with a fixed maximum length) between two CORE AS.
+Every AS forwards any received beacons to all neighbor CORE ASes unless this would cause a loop or exceed the fixed maximum length.
+Implemented naively, the number of paths (and beacons) grows exponentially with the network size.
+This is currently mitigated, primarily (and efficiently), by forwarding only the five "best PCBs" to neighbours.
+
+This "best PCB" practice has several implications:
+
+* It limits the number of available paths for endpoints such that only five truly diverse paths may be available to a given destination.
```
Comment (juagargi) [inline draft-meynell-panrg-scion-research-questions.md:213]: I don't agree with this. It might limit the number of paths if and only if there are no _new_ beacons arriving to the neighboring ASes after **N** beaconing periods. For that to happen, we require that the beaconing step _i+1_ forwards the same beacons than step _i_. OR that the beacons are very short lived and cannot survive longer than one period.
In any other case, period _i+1_ could, and would, forward different beacons to the neighboring ASes.

This is how I would explain it.
- The time needed to propagate all the beacons to the neighboring ASes increments.
- Since all beacons have an expiration time, we have a finite amount of time until a beaconing period will need to forward an already forwarded beacon, just because it expired.
- And, _most importantly_, the semantics of what a "best PCB" means, are defined locally by the **beacon forwarder AS**, not by the receiver AS, who is the actual judge of the quality of the PCB.

File: draft-meynell-panrg-scion-research-questions.md (hunk 4 above)
Comment (juagargi) [reply draft-meynell-panrg-scion-research-questions.md:220]: I agree with @tzaeschke _iff_ they forward the same PCBs during each beaconing period.
Let me give an example to clarify:
- First AS "A" discriminates good PCBs using latency criteria. This means that for AS "A" only good latency PCBs are forwarded.
- Second AS "B" receives PCBs from AS "A", but "B" discriminates bases on bandwidth. Maybe some PCBs have a good bandwidth, but many were already filtered by "A".
- etc.

If ASes forward _all received beacons_ (in general, and assuming there is no expiration time inside the PCBs), there should not be a problem, since all paths will be eventually discovered.

Now, because reality is that ASes cannot forward all of the PCBs, the non forwarded PCBs (those that an AS never forwards) are the least interesting ones _according to the local criteria of that AS_, which may or may not coincide with that criteria of the next neighbors this PCB will encounter.

I believe this to be clearer if explained in terms of completeness of the graph exploration vs. time needed for it.

File: draft-meynell-panrg-scion-research-questions.md [hunk 9]
```diff
@@ -201,11 +201,36 @@ This draft assumes the reader is familiar with some of the fundamental concepts
...
+  "truly diverse" means: Most ASes will receive five PCBs to a given destination from several of their neighbours. However, assuming
+  those ASes are use a similar PCB selection policy, it is not unlikely that each neighbor AS will forward a very
+  similar set of five PCB. In effect, there may be more than five paths available, but they are likely to diverge only on the first
+  one or two hops **(TBC: Is that true? There is no research confirming this hypothesis)**.
+  The problem that there may be only five truly diverse paths available is especially relevant for:
+  * Multipathing: endpoints may not be able to use more than five paths and no great selection of pats on offer.
+  * Unusual path policies, such as geofencing, may not be fulfillable by the limited number of paths on offer.
+* It causes a small number of paths to carry all traffic to a given destination, see {{link-load-balancing}}.
```
Comment (juagargi) [inline draft-meynell-panrg-scion-research-questions.md:221]: See comments above: only if the exploration achieved by this beaconing is very far from a full exploration of the complete graph, when the equality operator here `link1 = link2` means they have the same four tuple `[AS1,interface number1, AS2, interface number2]`

File: draft-meynell-panrg-scion-research-questions.md [hunk 10]
```diff
@@ -201,11 +201,36 @@ This draft assumes the reader is familiar with some of the fundamental concepts
...
+* It opens the possibility of a DoS attack where a small number of "wormholes", see {{dos-with-wormholes}}.
```
Comment (juagargi) [inline draft-meynell-panrg-scion-research-questions.md:222]: The possibility of attacking the selection of the best paths does not appear due to this beaconing process.
The same issue appears with an e.g. push-pull-publish-subscribe method, as long as the information propagated in the PCBs is trusted by the AS selecting the beacons and paths. Be that information latency, bandwidth, number of hops, etc (anything).

File: draft-meynell-panrg-scion-research-questions.md [hunk 11]
```diff
@@ -201,11 +201,36 @@ This draft assumes the reader is familiar with some of the fundamental concepts
...
+
+
 ## Segment Dissemination
 
-Control servers return a large number of path segments. This can cost considerable bandwidth / network egress while at the same time overloading clients with an unnecessarily large numbers of segments, mostly consisting of redundant information in terms of duplicate link and hops.
+Control servers return a large number of path segments when an endpoint requests paths to a destination.
```
Comment (juagargi) [inline draft-meynell-panrg-scion-research-questions.md:227]: Maybe refresh the process of an endpoint requesting a path from itself as SRC, to a given DST.
This will make clear how the core ASes' control service is contacted, how many times, etc.

File: draft-meynell-panrg-scion-research-questions.md [hunk 12]
```diff
@@ -201,11 +201,36 @@ This draft assumes the reader is familiar with some of the fundamental concepts
...
+This can cost considerable bandwidth / network egress while at the same time
+overloading clients with an unnecessarily large numbers of segments,
+mostly consisting of redundant information in terms of duplicate links and hops.
```
Comment (juagargi) [inline draft-meynell-panrg-scion-research-questions.md:230]: This is not immediately clear to me: why _mostly_ redundant information?

File: draft-meynell-panrg-scion-research-questions.md [hunk 13]
```diff
@@ -201,11 +201,36 @@ This draft assumes the reader is familiar with some of the fundamental concepts
...
 
-* This problem may be more problematic in ASes with many end hosts (e.g. IoT), or end hosts with little computing power or little spare bandwidth.
+* This problem may be more acute in ASes with many end hosts (e.g. IoT),
+or end hosts with little computing power or little spare bandwidth.
 * Getting a full path to a remote endhost may require three round-trips with the control server.
```
Comment (juagargi) [inline draft-meynell-panrg-scion-research-questions.md:234]: Three round trips from the local control service to control services of other two ASes.

File: draft-meynell-panrg-scion-research-questions.md [hunk 14]
```diff
@@ -214,9 +239,14 @@ There are multiple possible and independent solution steps here:
 * Allow queries from start AS to end AS across multiple segments. This should be very easy to implement and would be compatible with the current wire protocol (protobuf).
```
Comment (juagargi) [inline draft-meynell-panrg-scion-research-questions.md:239]: I don't understand the approach. Can you clarify?

File: draft-meynell-panrg-scion-research-questions.md [hunk 15]
```diff
@@ -303,6 +333,41 @@ Do we actually need to solve this reverse path refresh problem?
...
+A lack of recovery can lead to an attack as described in {{dos-with-wormholes}}.
+
+### Mitigation
+
+* An AS could monitor traffic for SCMP errors, however this works only if these errors are actually generated and forwarded.
+* Endpoints need a way to signal faulty segments to ASes. ASes need a way to signal faulty ASes to other ASes (e.g.
+core ASes or ASes in other ISDs).
+* ASes need adaptive becaon analysis algorithms that allow excluding specific beacons, e.g. beacons that have
+  known malicious ASes on path.
+* For swift recovery, it would be useful if ASes could revoke faulty segments.
```
Comment (juagargi) [inline draft-meynell-panrg-scion-research-questions.md:366]: Although we could revisit the topic, path revocation was studied already, and brought only more problems into the picture. For a full rationale, see the SCION book v2, 5.7 Path Revokation

File: draft-meynell-panrg-scion-research-questions.md [hunk 16]
```diff
@@ -303,6 +333,41 @@ Do we actually need to solve this reverse path refresh problem?
...
+* Adapt the path discovery and dissemination algorithm to allow ASes to have many more than five paths to
```
Comment (juagargi) [inline draft-meynell-panrg-scion-research-questions.md:367]: As a comment: This is essentially what the book promises, that the endpoints will have anyway a big selection to choose paths from.

File: draft-meynell-panrg-scion-research-questions.md [hunk 17]
```diff
@@ -328,6 +393,30 @@ To be discussed
...
+see {{I-D.scion-cp}}, between to "major" locations, e.g. between Europe and Australia.
+The five wormholes advertise excellent traffic conditions by advertising low latency, few hops, low cost, ...).
+As a result the beacons of the five wormholes are likely to be chosen as the "five best" segments for any traffic
+between Europe and Australia. After all alternative paths (which may be in local oath databases) have expired,
+the wormholes will be the only paths available to local ASes.
+
+### Impact
+The attacker can now perform blackhole attacks (DoS), greyhole attacks, delay traffic, or similarily impact traffic.
+
+Critically, there is no way no recover from this situation. In order to recover, ASes that are connected to the
```
Comment (juagargi) [inline draft-meynell-panrg-scion-research-questions.md:408]:
```suggestion
Critically, there is no current way to recover from this situation. In order to recover, ASes that are connected to the
```

**Review by juagargi — CHANGES_REQUESTED** [review]

File: draft-meynell-panrg-scion-research-questions.md (hunk 15 above)
Comment (tzaeschke) [reply draft-meynell-panrg-scion-research-questions.md:366]: Thanks for the pointer. It turns out, path revocation comes basically for free with 2-phase beaconing :-D.

Comment (juagargi) [reply draft-meynell-panrg-scion-research-questions.md:366]: Is there a concept of emitting information to invalidate a previously authorized PCB?
If so, that is what we call a revocation. If it is not possible to invalidate previously authorized PCB, then I would call it differently, to distinguish from _revocation_ used in the book or previous I-Ds.

**Review by juagargi — CHANGES_REQUESTED** [review]

File: draft-meynell-panrg-scion-research-questions.md (hunk 7 above)
Comment (tzaeschke) [reply draft-meynell-panrg-scion-research-questions.md:216]: Just to clarify the example, we are looking for paths from X to Y:
```
1) Y->a->b->c -> X
2) Y->a'->b'->c' -> X
```
At a first step I'd argue that it is quite likely that, going from Y->X,  we have a'=a. If a=a' then the same argument applies to b=b' and so forth.
Consequently I'd argue that there is a high probability of overlap.

Unfortunately, I don't think there is any analysis or experiments to confirm/reject this. If people think this is wrong then we can remove it. Or I can try to rephrase it as a research question that would need to be clarified whether this may be a problem or not?

File: draft-meynell-panrg-scion-research-questions.md (hunk 8 above)
Comment (tzaeschke) [reply draft-meynell-panrg-scion-research-questions.md:213]: Sorry, I don't understand what you are saying. Maybe you can clarify?

My assumption is that an AS will rarely change it's what-is-the-best-PCB algorithm. Consequently, they would always forward the same five segments. They may receive new beacons, but as long as the "best 5" are still among them (and I assume they will be, why shouldn'´t they?), they will continue forwarding these 5 PCBs.

File: draft-meynell-panrg-scion-research-questions.md (hunk 13 above)
Comment (tzaeschke) [reply draft-meynell-panrg-scion-research-questions.md:234]: I actually meant from the local endpoint to the local control server. I am less concerned about roundtrips between ASes, they have stonger machines and better connections.

File: draft-meynell-panrg-scion-research-questions.md (hunk 14 above)
Comment (tzaeschke) [reply draft-meynell-panrg-scion-research-questions.md:239]: I did. Can you have another look?

File: draft-meynell-panrg-scion-research-questions.md (hunk 16 above)
Comment (tzaeschke) [reply draft-meynell-panrg-scion-research-questions.md:367]: Done.

File: draft-meynell-panrg-scion-research-questions.md (hunk 13 above)
Comment (juagargi) [reply draft-meynell-panrg-scion-research-questions.md:267]: AFAIK the endpoint requests once the path to the destination, and receives up to three segments, all in one round trip.
The CS of the local AS may have to do several trips to core ASes in its local ISD, and then core ASes in the remote ISD.
But I think the general idea is understood, approving.

**Review by juagargi — APPROVED** [review]
